src/dependency-manager/src/configs/graph/index.ts

```diff
@@ -0,0 +1,262 @@
+      this._enriched_config.setServices(enriched_services);
+    }
+
+    // Assign hosts/ports to interfaces as needed
+    const interfaces = config.getInterfaces();
+    for (const [key, value] of interfaces) {
+      value.host = value.host || await options.getHostAssignment(config, key);
+      value.port = value.port || await options.getPortAssignment(config, key);
+      interfaces.set(key, value);
+    }
+    config.setInterfaces(interfaces);
+    this._enriched_config.addService(config);
+
+    // Assert that we've already enriched this service
+    options.skip_nodes = options.skip_nodes || [];
+    options.skip_nodes.push(config.getName() || '');
+
+    // Recurse through dependencies
+    for (const dependency of config.getDependencies()) {
+      await this.buildService(dependency, options);
+    }
+
+    // Load parameters (must happen after dependencies are loaded so value_from
+    // references will work)
+    config = await this.loadParametersForService(config);
+    this._enriched_config.addService(config);
+  }
+
+  private async build(
+    options: GraphEnrichmentOptions,
+    force_rebuild = false,
+  ) {
+    if (!this._is_built || force_rebuild) {
+      this._enriched_config = EnvironmentBuilder.create();
+
+      options = {
+        debug: false,
+        ...options,
+      };
+
+      for (const service of this._base_config.getServices()) {
+        await this.buildService(service, options);
+      }
+
+      this._is_built = true;
+    }
+  }
+
+  /**
+   * Build out the enriched environment config from the base config to prep
+   * for deployment
+   */
+  static async build(
+    config: BaseEnvironmentConfig,
+    options: GraphEnrichmentOptions
+  ): Promise<EnvironmentGraph> {
+    const graph = new EnvironmentGraph(config);
+    await graph.build(options);
+    return graph;
+  }
+
+  /**
+   * Run validation checks on the graph to ensure it has everything it needs
+   * in order to be deployed. This should not make network calls and should
+   * just check the contents of the _enriched_config
+   * @returns boolean
+   */
+  validate(): boolean {
+    throw new Error('Run validation checks');
+  }
+
+  /**
+   * Retrieve the list of services in the graph. Will throw an error if the full
+   * list hasn't been built out
+   */
+  getServices() {
+    if (!this._enriched_config) {
+      throw new GraphNotBuiltError();
+    }
+
+    return this._enriched_config.getServices();
+  }
+
+  /**
+   * Add/update a service in the graph with a new config. Will throw an error if
+   * the full graph hasn't been built out.
+   * @param service [BaseServiceConfig]
+   */
+  addService(service: BaseServiceConfig) {
+    if (!this._enriched_config) {
+      throw new GraphNotBuiltError();
+    }
+
+    this._enriched_config.addService(service);
+    this._base_config.addService(service);
+  }
+
+  /**
+   * Retrieve the vault instances used to interact with secret managers
+   */
+  getVaultManager(): VaultManager {
```

Comment:
Does this get used? We don't want to create multiple versions of the VaultManager because it caches results.